File: 2542. Maximum Subsequence Score/Solution.java
class Solution {
    public long maxScore(int[] nums1, int[] nums2, int k) {
        int n = nums1.length, i;
        long total = 0, res = Long.MIN_VALUE;
        int[][] combined = new int[n][2];
        PriorityQueue<Integer> q = new PriorityQueue<>();
        for (i = 0; i < n; i++) {
            combined[i][0] = nums1[i];
            combined[i][1] = nums2[i];
        }
        Arrays.sort(combined, (a, b) -> {
            if (a[1] == b[1]) return b[0] - a[0];
            return b[1] - a[1];
        });
        for (i = 0; i < n; i++) {
            q.offer(combined[i][0]);
            total += combined[i][0];
            if (i >= k) total -= q.poll();
            if (i >= k - 1) res = Math.max(res, total * combined[i][1]);
        }
        return res;
    }
}

File: 2542. Maximum Subsequence Score/Solution.cpp
class Solution {
public:
    long long maxScore(vector<int>& nums1, vector<int>& nums2, int k) {
        int n = nums1.size();
        long long total = 0, res = 0;
        vector<pair<int, int>> combined;
        priority_queue<int, vector<int>, greater<int>> pq;
        for (int i = 0; i < n; i++) {
            pair<int, int> p(nums1[i], nums2[i]);
            combined.push_back(p);
        }
        auto cmp = [](auto& a, auto& b) -> bool { return (a.second == b.second) ? b.first < a.first : b.second < a.second; };
        sort(begin(combined), end(combined), cmp);
        for (auto& c : combined) {
            pq.push(c.first);
            total += c.first;
            if (pq.size() == k + 1) {
                total -= pq.top();
                pq.pop();
            }
            if (pq.size() == k) {
                res = max(res, total * c.second);
            }
        }
        return res;
    }
};
